File: lib/entity/fetch_status.dart
import "package:json_annotation/json_annotation.dart";

part "fetch_status.g.dart";

@JsonSerializable()
class FetchStatus {
  FetchAction action;
  List<String> args;
  int progress;
  int max;
  FetchStatus({
    required this.action,
    required this.args,
    required this.progress,
    required this.max,
  });

  factory FetchStatus.fromJson(Map<String, dynamic> json) {
    return _$FetchStatusFromJson(json);
  }

  Map<String, dynamic> toJson() {
    return _$FetchStatusToJson(this);
  }

  @override
  String toString() {
    return 'FetchStatus(action: $action, args: $args, progress: $progress, max: $max)';
  }
}

enum FetchAction {
  fetchConfiguration,
  fetchProviders,
  verifying,
}

File: ios/Classes/entity/FetchStatus.swift
//
//  FetchStatus.swift
//  clash_flt
//
//  Created by LondonX on 2022/8/26.
//

import Foundation

class FetchStatus : NSObject{
    let action: Action
    let args: [String]
    let progress: Int
    let max: Int
    init(
        action: Action,
        args: [String] = [],
        progress: Int = 0xFFFF,
        max: Int = 0xFFFF
    ){
        self.action = action
        self.args = args
        self.progress = progress
        self.max = max
    }
    
    func toMap() -> [String : Any] {
        return [
            "action" : action.rawValue,
            "args" : args,
            "progress" : progress,
            "max" : max,
        ]
    }
}

enum Action: String{
    case fetchConfiguration
    case fetchProviders
    case verifying
}
